FC-885 add validation for basket save in favourites api

ben1stone-leftdfe opened the pull request.

ben1stone-leftdfe committed:
Reorganising

ben1stone-leftdfe committed:
Added unit tests for returning bad request when an ApprenticeshipId or UKprn is invalid.

leewadhams commented on src/api/EmployerFavouritesApi/EmployerFavouritesApi.csproj:
```diff
@@ -35,4 +35,6 @@
     </Content>
   </ItemGroup>
 
+  <ProjectExtensions><VisualStudio><UserProperties appsettings_1development_1json__JsonSchema="" /></VisualStudio></ProjectExtensions>
```
What is this?

leewadhams commented on src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs:
```diff
@@ -100,11 +106,40 @@ public async Task<IActionResult> Put(string employerAccountId, [FromBody]List<Fa
 
                 return NoContent();
             }
+            catch (InvalidUkprnException ex)
```
Do we need custom exceptions or would ArgumentException suffice?

leewadhams commented on src/api/EmployerFavouritesApi/appsettings.json:
```diff
@@ -21,7 +21,7 @@
   "ConnectionStrings": {
     "Redis": "",
-    "AzureStorage": ""
+    "AzureStorage": "UseDevelopmentStorage=true"
```
Don't think this should be in this file.

leewadhams commented on src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs:
```diff
@@ -100,11 +106,40 @@ public async Task<IActionResult> Put(string employerAccountId, [FromBody]List<Fa
+            }
+            catch (InvalidApprenticeshipIdException ex)
+            {
+                return BadRequest(ex);
```
Even though it's an internal api we should probably return the message not the exception so that the internals of the app aren't leaked e.g. stack traces.
```csharp
            {       
                return BadRequest(ex.Message);              
            }
```

leewadhams commented on src/api/EmployerFavouritesApi/Validation/ApprenticeshipsParameterValidator.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    internal class ApprenticeshipsParameterValidator
+    {
+        private Regex _frameworkRegex = new Regex(@"^\d{3}-\d{1,2}-\d{1,2}$");
```
could be made readonly

leewadhams commented on src/api/EmployerFavouritesApi/Validation/ApprenticeshipsParameterValidator.cs:
```diff
@@ -0,0 +1,40 @@
+            return false;
+        }
+
+        internal bool IsValidEmployerAccountId(string employerAccountId)
```
This isn't used. Are we missing applying it on the controller?

leewadhams reviewed (changes requested):
Nice. Couple of small changes.

ben1stone-leftdfe committed:
Updates after code review
- Replaced custom exceptions with ArguementException
- Implemented the IsValidEmployerAccountId paramValidator
- Removed development settings from appsettings.json
- Returned the message on BadREquest not the whole exception
- Improved logging

sultansfa1 reviewed (approved):
This has been tested successfully and PO reviewed
